Cache unique values when creating features

elpaso opened the pull request:
Cache unique values when creating features

Fixes #21305 - pasting features is very slow

Aggressively optimize createFeature for speed
and introduces createFeatures for bulk creation.

elpaso committed:
Cache unique values in createFeature (~10x faster)

elpaso committed:
Cache unique values when creating features

Fixes #21305 - pasting features is very slow

Aggressively optimize createFeature for speed
and introduces createFeatures for bulk creation.

elpaso commented:
@nyalldawson @m-kuhn I'd appreciate your review on this one.

Speed gain is really huge and it makes pasting features usable again.

elpaso committed:
Rename QgsFeaturesData to QgsFeatureData and fix test case

elpaso committed:
Fix docs  for createFeatures

m-kuhn commented on src/core/qgsvectorlayerutils.h:
```diff
@@ -98,8 +98,8 @@ class CORE_EXPORT QgsVectorLayerUtils
     };
 
     // SIP does not lile "using", use legacy typedef
```
Typo: like (I know it's not your fault ;P )

m-kuhn commented:
I like the idea

The only thing I'm not sure about is if it's worth mentioning a performance factor, there's probably a linear dependency between the speed gain and the number of features :)

elpaso commented:
@m-kuhn thanks for the review, how do you feel about backporting, is that ok?

m-kuhn commented:
I would give it a considerable amount of testing time before backporting given that there's a risk of side effects.

gioman commented:
> I would give it a considerable amount of testing time before backporting given that there's a risk of side effects.

@m-kuhn I will do my best to test it, anyway this possibly solve a very bad regression (and hopefully also fixes the other issue about opening time of attribute tables with many columns, that is QGIS 3 has become unbearably slow).

elpaso committed:
Typo

elpaso committed:
Fix unique values on GPKG pk

Fixes #21311

elpaso committed:
Fix renamed QgsVectorLayerUtils.QgsFeatureData

elpaso committed:
Fix unique values on SQLite (OGR) pk

elpaso committed:
Explain why the workaround for GPKG/SQLite is necessary

I've talked to Even for this one

m-kuhn commented on src/providers/ogr/qgsogrprovider.cpp:
```diff
@@ -3618,7 +3618,19 @@ QSet<QVariant> QgsOgrProvider::uniqueValues( int index, int limit ) const
   QByteArray sql = "SELECT DISTINCT " + quotedIdentifier( textEncoding()->fromUnicode( fld.name() ) );
+
+  // GPKG/SQLite fid
+  // For GPKG an SQLITE drivers PK fields are not exposed as real fields, (and OGR_F_GetFID only
```
```suggestion
  // For GPKG and SQLITE drivers PK fields are not exposed as real fields, (and OGR_F_GetFID only
```

m-kuhn committed:
Update src/providers/ogr/qgsogrprovider.cpp

Co-Authored-By: elpaso <elpaso@itopen.it>

m-kuhn commented on src/providers/ogr/qgsogrprovider.cpp:
```diff
@@ -3618,7 +3618,19 @@ QSet<QVariant> QgsOgrProvider::uniqueValues( int index, int limit ) const
+  if ( ( mGDALDriverName == QLatin1String( "GPKG" ) || mGDALDriverName == QLatin1String( "SQLite" ) )
+       && mFirstFieldIsFid && index == 0 )
+  {
+    sql += ", " + quotedIdentifier( textEncoding()->fromUnicode( fld.name() ) ) + " AS fid2";
```
```suggestion
    sql += QStringLiteral( ", " ) + quotedIdentifier( textEncoding()->fromUnicode( fld.name() ) ) + QStringLiteral( " AS fid2" );
```

m-kuhn committed:
Update src/providers/ogr/qgsogrprovider.cpp

Co-Authored-By: elpaso <elpaso@itopen.it>

elpaso commented:
@m-kuhn sql is QByteArray

m-kuhn commented:
oops

elpaso committed:
Revert QStringLiteral

m-kuhn commented:
What I wonder more is if it's appropriate to trigger a `createFeature()` on every paste operation.

Shouldn't the logic be

 * If a pasted feature misses an attribute (really misses, not if it's NULL) then get a default value for this attribute
 * Then check unique constraints and if some unique constraints fail, try to get a unique attribute

Or is this not the same kind of pasting we are talking about?

elpaso commented:
> What I wonder more is if it's appropriate to trigger a `createFeature()` on every paste operation.
> 
> Shouldn't the logic be
> 
>     * If a pasted feature misses an attribute (really misses, not if it's NULL) then get a default value for this attribute
> 
>     * Then check unique constraints and if some unique constraints fail, try to get a unique attribute
> 
> 
> Or is this not the same kind of pasting we are talking about?

I'm not sure I follow you: this is exactly what happens inside the monster method `createFeature/s`: all constraints are checked, all defaults are evaluated and the last step tries to create a unique value if the unique constraint is still not honored.
 
I'm currently working on a fix for this last step: we should use the cached unique values instead of fetching from the provider (because the provider does not have all the values from the possibly previously other created features).

Btw, I'm happy to talk/chat to you about this: maybe easier to explain.

elpaso committed:
Fix unique values when generating a set of features

elpaso commented:
@m-kuhn see my last commit for the unique values generation (now using the cached unique values) and the new test that failed with the previous implementation.

gioman commented:
I have made some test, and it seems that the "slow paste" bug is fixed (saves  after that still can be very slow, but this is a different long standing issue).

On the other hand https://issues.qgis.org/issues/21303 seems not fixed by this patch: the same table with ~60 columns and ~4k records takes 5/7 seconds to open in 2.18, while is well over a minute in 3.4.4/master.

elpaso commented:
@m-kuhn good to merge or do you want to review this deeper?

m-kuhn commented on src/core/qgsvectorlayerutils.cpp:
```diff
@@ -375,88 +463,104 @@ QgsFeature QgsVectorLayerUtils::createFeature( const QgsVectorLayer *layer, cons
       {
-        v = providerDefault;
-        checkUnique = false;
+        v = fd.attributes().value( idx );
```
```suggestion
        v = fd.attributes().at( idx );
```

The entry is guaranteed to exist at this point

m-kuhn replied:
Should there also be a null substitution for invalid values in here?

```
if ( !v.isValid() )
  v = QVariant( QVariant::Integer );
```

I am quite sure some pasted features will have invalid variants in the place of NULL variants which is incompatible with https://github.com/qgis/QGIS/pull/9203/files#diff-ad7e488f21bc920cdf5407f3b17f5c42R497

I'm not sure if it's better to not do it and fix those down the road (but I guess they will be hard to catch and lead to obscure behaviour)
Or if we should play it safe in code like this and at the same time slowly try to bring the codebase to a safe null vs. invalid condition.

elpaso replied:
@m-kuhn 
./src/core/qgsvectorlayerutils.cpp:488:29: error: no member named 'at' in 'QMap<int, QVariant>'
        v = fd.attributes().at( idx );

m-kuhn replied:
Ok, leave as is then!

m-kuhn commented on src/core/qgsvectorlayerutils.cpp:
```diff
@@ -375,88 +463,104 @@ QgsFeature QgsVectorLayerUtils::createFeature( const QgsVectorLayer *layer, cons
-      v = attributes.value( idx );
-    }
+      // 3. provider side default value clause
+      // note - not an else if deliberately. Users may return null from a default value expression to fallback to provider defaults
```
Do we really want that? NULL to fallback to provider defaults... different discussion.

m-kuhn replied:
Actually the check below is for `isValid()` and the comment here says NULL

elpaso replied:
Yeah, different discussion (that we need to do sooner or later), I called this `monster ` method for a reason, I don't entirely get the logic here since there are a lot of different permutations of use cases.

m-kuhn commented:
@elpaso I see how it works now, looks reasonable. Very good!
I just left some comments which I think already applied to the code before, would be good to get @nyalldawson 's eyes there too, if I'm not mistaken he's the original author of that.

m-kuhn committed:
Update src/core/qgsvectorlayerutils.cpp

Co-Authored-By: elpaso <elpaso@itopen.it>

nirvn commented:
Is the plan for this to be merged into 3.6.0, or is it better to aim at 3.6.1 to give it more testing time?

elpaso commented:
@nirvn I'd merge into 3.6: there is quite a good test coverage on this area and without this patch pasting a bunch of features with a "slow" provider makes QGIS completely unresponsive for minutes.

I removed the backporting tag, but if it was my decision I would backport it.

m-kuhn commented:
I agree, merging for 3.6.0 is a good idea, it looks "safe enough" and more testing is welcome. For backporting I think the question is not *if* but *when*.

nirvn commented:
Ok. If you guys think it's safe enough, I'm happy with that :)

elpaso committed:
Revert "Update src/core/qgsvectorlayerutils.cpp"

This reverts commit f4d7e73a545dcf2604b93f55f5b0030a5a29440a.

elpaso committed:
Moved comment

elpaso committed:
Prevent crash when fields are deleted from the vector options dialog

elpaso committed:
Also check for nulls when applying defaults

isValid is not enough because fields are initialized with
QVariant(field.type()) which is valid but null.

Fixes #21304

With test

elpaso commented:
@m-kuhn is this what you meant with checking for nulls? https://github.com/qgis/QGIS/pull/9203/commits/bf1575f4c9aca6c550f6d0711bd3297adfbb27b2

I added a test for bug https://issues.qgis.org/issues/21304

elpaso committed:
Test for regression #21304 "default value" is not applied when pasting features

m-kuhn commented:
Codewise, `v.isNull()` implicitly also checks for `!v.isValid()` so the extra check there is no longer required if this change is done.

However, I was more questioning this functionality in general. I.e. I believe a QGIS configured default value should be able to overwrite a database default value with a `NULL` value.

Maybe we need to add a new `INVALID` value that can be used by default values so we no longer abuse `NULL`.

But I'd really appreciate a statement by @nyalldawson concerning this before deciding for a path.

elpaso committed:
Removed redundant check for isValid

gioman commented:
> For backporting I think the question is not _if_ but _when_.

so this will not be backported? I understand playing safe but please also consider the bad regression (for a pretty common operation) we will eventually ship into the release that will put to bed 2.18.

m-kuhn commented:
> so this will not be backported?

on the contrary, it will be backported when we have enough test results to be sure it doesn't introduce any regressions.

gioman commented:
> on the contrary,

sorry I expressed myself badly... I was thinking (in time) for the upcoming release.

m-kuhn commented:
@gioman @elpaso do you feel confident for an express backport on this?
I'm not putting a veto on it, just thought another couple of weeks for testing this might be good, but if you feel it's appropriate to move ahead, please proceed.

nyalldawson commented:
My vote would go to playing it safe and waiting till 3.4.6 here -- I understand it's a major regression, but this code is quite complex and there's a chance a worse regression may have been introduced!

gioman commented:
> My vote would go to playing it safe and waiting till 3.4.6 here -- I understand it's a major regression, but this code is quite complex and there's a chance a worse regression may have been introduced!

@nyalldawson @elpaso @m-kuhn I understand, no worries. Thanks everyone for the great bug fixing effort!
